File: packages/admin/ts/src/control.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Control API client that mints a scoped, single-use Caracal token per call and invokes a control command through the governed /v1/control/invoke path.

const STS_TOKEN_PATH = '/oauth/2/token'
const CONTROL_INVOKE_PATH = '/v1/control/invoke'
const DEFAULT_TIMEOUT_MS = 30_000

// The control identity the caller acts as. The client secret is a sealed credential held
// only for the lifetime of a request and never logged; it leaves this module only in the
// token-exchange request body to the STS.
export interface ControlClientOptions {
  stsUrl: string
  controlUrl: string
  audience: string
  applicationId: string
  clientSecret: string
  // The lifetime to request for each minted token. Tokens are single-use (replay
  // protected) so this only bounds the window between mint and invoke; a short value
  // keeps the blast radius of a leaked token minimal.
  ttlSeconds?: number
  // The zone a read invoke targets when it is not the identity's own zone. The token is still
  // minted in the identity's home zone; this rides as a header the control handler honors
  // only for a reserved reader identity and only for non-mutating commands, so such a reader
  // can inspect a tenant zone's live state without provisioning any identity in that zone.
  // Absent for an identity acting in its own zone, which is the common path.
  zoneScope?: string
  // An optional attribution string the acting identity asserts for the human or upstream
  // authority on whose behalf it invokes. It rides in the invoke body and is recorded verbatim
  // in the control.invoke audit metadata as a subject-asserted annotation; it is never an
  // authorization input, so the action stays bounded entirely by the token's own scopes and zone.
  authorizedBy?: string
  // Marks the invoke as originating from the Caracal Operator, so an object it creates is stamped
  // with operator co-authorship. Set only by the Operator's own governed execution path, never by
  // direct control-plane automation.
  coAuthorOperator?: boolean
  // The originating request id, forwarded as x-request-id on both the token exchange and the
  // control invoke so the STS log, the control dispatch, and its audit record all correlate
  // back to the request that triggered them.
  requestId?: string
  // Total wall-clock budget for one invoke, including token mint, its one safe retry,
  // and the control request. Caller cancellation can shorten this budget.
  timeoutMs?: number
  signal?: AbortSignal
  fetchImpl?: typeof fetch
}

// A control invoke failed. stage distinguishes a token-exchange failure from a control
// dispatch failure; status 0 means the request itself failed and no response arrived.
// reason is already free of the client secret, so it is safe to surface or log. code and
// remediation are the structured control-plane fields when the failure came from dispatch.
export class ControlClientError extends Error {
  constructor(
    public readonly stage: 'token' | 'invoke',
    public readonly status: number,
    public readonly reason: string,
    public readonly code?: string,
    public readonly remediation?: string,
  ) {
    super(`control ${stage} failed (${status}): ${reason}`)
    this.name = 'ControlClientError'
  }

  // Whether the failure provably applied nothing: any token-stage failure (no token was
  // minted, so nothing was invoked) or an invoke the control plane rejected with a client
  // error. An invoke-stage server error or lost response is not definitive - the command
  // may already have applied - so a caller must never blindly retry it.
  get definitive(): boolean {
    return this.stage === 'token' || (this.status >= 400 && this.status < 500)
  }
}

interface ControlErrorBody {
  error?: { code?: string; reason?: string; remediation?: string } | string
}

function describeError(body: unknown, fallback: string): { reason: string; code?: string; remediation?: string } {
  const envelope = (body as ControlErrorBody | null)?.error
  if (envelope && typeof envelope === 'object') {
    return { reason: envelope.reason ?? fallback, code: envelope.code, remediation: envelope.remediation }
  }
  return { reason: typeof envelope === 'string' ? envelope : fallback }
}

async function readJson(res: Response): Promise<unknown> {
  const text = await res.text()
  if (text.length === 0) return undefined
  try {
    return JSON.parse(text)
  } catch {
    return { raw: text }
  }
}

// Trims trailing slashes with a linear scan: a quantified-regex trim backtracks
// quadratically on long slash runs (js/polynomial-redos), and the URL is library input.
function trimTrailingSlashes(url: string): string {
  let end = url.length
  while (end > 0 && url[end - 1] === '/') end--
  return url.slice(0, end)
}

// A control-plane client bound to one identity. Each invoke mints a fresh token scoped to
// exactly the scopes that call requires, so an action carries the least authority that
// satisfies it and a leaked token grants nothing beyond that one operation. The caller
// decides what to invoke and with which scopes; this client only carries it to the
// governed control surface.
export class ControlClient {
  private readonly stsUrl: string
  private readonly controlUrl: string
  private readonly fetchImpl: typeof fetch

  constructor(private readonly options: ControlClientOptions) {
    this.stsUrl = trimTrailingSlashes(options.stsUrl)
    this.controlUrl = trimTrailingSlashes(options.controlUrl)
    this.fetchImpl = options.fetchImpl ?? fetch
  }

  // Exchanges the identity's client credentials for a Caracal control token scoped to
  // exactly the requested scopes. The STS narrows the token to the intersection of the
  // identity's allowed scopes and those requested, so an over-broad request can never
  // widen authority beyond what the identity was granted. A transient failure (a server
  // error or a lost response) is retried once: a failed mint is always definitive - no
  // token exists and nothing was applied - so the retry is safe for every caller.
  private async mintToken(scopes: readonly string[], signal: AbortSignal): Promise<string> {
    try {
      return await this.exchangeToken(scopes, signal)
    } catch (err) {
      if (err instanceof ControlClientError && (err.status >= 500 || err.status === 0)) {
        return this.exchangeToken(scopes, signal)
      }
      throw err
    }
  }

  private async exchangeToken(scopes: readonly string[], signal: AbortSignal): Promise<string> {
    const form = new URLSearchParams({
      grant_type: 'client_credentials',
      application_id: this.options.applicationId,
      client_secret: this.options.clientSecret,
      resource: this.options.audience,
      scope: scopes.join(' '),
    })
    if (this.options.ttlSeconds !== undefined) form.set('ttl_seconds', String(this.options.ttlSeconds))
    const headers: Record<string, string> = { 'content-type': 'application/x-www-form-urlencoded' }
    if (this.options.requestId) headers['x-request-id'] = this.options.requestId
    const res = await this.send(
      'token',
      `${this.stsUrl}${STS_TOKEN_PATH}`,
      {
        method: 'POST',
        headers,
        body: form.toString(),
      },
      signal,
    )
    const body = await readJson(res)
    if (!res.ok) {
      const { reason, code, remediation } = describeError(body, 'token exchange rejected')
      throw new ControlClientError('token', res.status, reason, code, remediation)
    }
    const token = (body as { access_token?: unknown } | undefined)?.access_token
    if (typeof token !== 'string' || token.length === 0) {
      throw new ControlClientError('token', res.status, 'token exchange returned no access_token')
    }
    return token
  }

  // Carries a request to the wire, normalizing a thrown network failure into the error
  // taxonomy as status 0 so every failure a caller sees carries a stage and a status.
  private async send(stage: 'token' | 'invoke', url: string, init: RequestInit, signal: AbortSignal): Promise<Response> {
    try {
      return await this.fetchImpl(url, { ...init, signal })
    } catch (err) {
      throw new ControlClientError(stage, 0, err instanceof Error ? err.message : 'network request failed')
    }
  }

  async invoke(command: string, subcommand: string, flags: Record<string, unknown>, scopes: readonly string[]): Promise<unknown> {
    const timeoutMs = this.options.timeoutMs ?? DEFAULT_TIMEOUT_MS
    if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) throw new Error('ControlClient timeoutMs must be a positive number')
    const timeout = AbortSignal.timeout(timeoutMs)
    const signal = this.options.signal ? AbortSignal.any([timeout, this.options.signal]) : timeout
    const token = await this.mintToken(scopes, signal)
    const headers: Record<string, string> = { 'content-type': 'application/json', authorization: `Bearer ${token}` }
    if (this.options.zoneScope) headers['x-caracal-zone-scope'] = this.options.zoneScope
    if (this.options.requestId) headers['x-request-id'] = this.options.requestId
    const invokeBody: Record<string, unknown> = { command, subcommand, flags }
    if (this.options.authorizedBy) invokeBody.authorized_by = this.options.authorizedBy
    if (this.options.coAuthorOperator) invokeBody.co_author_operator = true
    const res = await this.send(
      'invoke',
      `${this.controlUrl}${CONTROL_INVOKE_PATH}`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify(invokeBody),
      },
      signal,
    )
    const body = await readJson(res)
    if (!res.ok) {
      const { reason, code, remediation } = describeError(body, 'control invoke rejected')
      throw new ControlClientError('invoke', res.status, reason, code, remediation)
    }
    return (body as { result?: unknown } | undefined)?.result
  }
}

File: packages/admin/go/control.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Control API client that mints a scoped, single-use Caracal token per call and invokes a control command through the governed /v1/control/invoke path.

package admin

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	stsTokenPath      = "/oauth/2/token"
	controlInvokePath = "/v1/control/invoke"
)

// ControlClientOptions configures a ControlClient bound to one identity.
type ControlClientOptions struct {
	STSURL           string
	ControlURL       string
	Audience         string
	ApplicationID    string
	ClientSecret     string
	TTLSeconds       int
	ZoneScope        string
	AuthorizedBy     string
	CoAuthorOperator bool
	RequestID        string
	HTTPClient       *http.Client
}

// ControlClientError is a failed control invoke. Stage distinguishes a
// token-exchange failure from a control dispatch failure; Status 0 means the
// request itself failed and no response arrived. Reason is already free of
// the client secret, so it is safe to surface or log. Code and Remediation
// are the structured control-plane fields when the failure came from
// dispatch.
type ControlClientError struct {
	Stage       string
	Status      int
	Reason      string
	Code        string
	Remediation string
}

func (e *ControlClientError) Error() string {
	return fmt.Sprintf("control %s failed (%d): %s", e.Stage, e.Status, e.Reason)
}

// Definitive reports whether the failure provably applied nothing: any
// token-stage failure (no token was minted, so nothing was invoked) or an
// invoke the control plane rejected with a client error. An invoke-stage
// server error or lost response is not definitive - the command may already
// have applied - so a caller must never blindly retry it.
func (e *ControlClientError) Definitive() bool {
	return e.Stage == "token" || (e.Status >= 400 && e.Status < 500)
}

// ControlClient is a control-plane client bound to one identity. Each invoke
// mints a fresh token scoped to exactly the scopes that call requires, so an
// action carries the least authority that satisfies it and a leaked token
// grants nothing beyond that one operation. The client secret is a sealed
// credential that leaves this package only in the token-exchange request body
// to the STS and is never logged.
type ControlClient struct {
	opts       ControlClientOptions
	httpClient *http.Client
}

// NewControlClient builds a ControlClient from options, trimming trailing
// slashes from the base URLs.
func NewControlClient(opts ControlClientOptions) *ControlClient {
	opts.STSURL = strings.TrimRight(opts.STSURL, "/")
	opts.ControlURL = strings.TrimRight(opts.ControlURL, "/")
	client := opts.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: 30 * time.Second}
	}
	return &ControlClient{opts: opts, httpClient: client}
}

// Invoke mints a token scoped to exactly the requested scopes and dispatches
// one control command, returning the command result.
func (c *ControlClient) Invoke(ctx context.Context, command, subcommand string, flags map[string]any, scopes []string) (any, error) {
	token, err := c.mintToken(ctx, scopes)
	if err != nil {
		return nil, err
	}
	body := map[string]any{"command": command, "subcommand": subcommand, "flags": flags}
	if c.opts.AuthorizedBy != "" {
		body["authorized_by"] = c.opts.AuthorizedBy
	}
	if c.opts.CoAuthorOperator {
		body["co_author_operator"] = true
	}
	payload, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.opts.ControlURL+controlInvokePath, strings.NewReader(string(payload)))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)
	if c.opts.ZoneScope != "" {
		req.Header.Set("X-Caracal-Zone-Scope", c.opts.ZoneScope)
	}
	if c.opts.RequestID != "" {
		req.Header.Set("X-Request-Id", c.opts.RequestID)
	}
	res, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &ControlClientError{Stage: "invoke", Status: 0, Reason: err.Error()}
	}
	defer res.Body.Close()
	parsed := readBody(res.Body)
	if res.StatusCode >= 400 {
		reason, code, remediation := describeError(parsed, "control invoke rejected")
		return nil, &ControlClientError{Stage: "invoke", Status: res.StatusCode, Reason: reason, Code: code, Remediation: remediation}
	}
	if envelope, ok := parsed.(map[string]any); ok {
		return envelope["result"], nil
	}
	return nil, nil
}

// mintToken exchanges the identity's client credentials for a control token
// scoped to exactly the requested scopes. A transient failure (a server error
// or a lost response) is retried once: a failed mint is always definitive -
// no token exists and nothing was applied - so the retry is safe for every
// caller.
func (c *ControlClient) mintToken(ctx context.Context, scopes []string) (string, error) {
	token, err := c.exchangeToken(ctx, scopes)
	if err == nil {
		return token, nil
	}
	var controlErr *ControlClientError
	if errors.As(err, &controlErr) && (controlErr.Status >= 500 || controlErr.Status == 0) {
		return c.exchangeToken(ctx, scopes)
	}
	return "", err
}

func (c *ControlClient) exchangeToken(ctx context.Context, scopes []string) (string, error) {
	form := url.Values{
		"grant_type":     {"client_credentials"},
		"application_id": {c.opts.ApplicationID},
		"client_secret":  {c.opts.ClientSecret},
		"resource":       {c.opts.Audience},
		"scope":          {strings.Join(scopes, " ")},
	}
	if c.opts.TTLSeconds > 0 {
		form.Set("ttl_seconds", strconv.Itoa(c.opts.TTLSeconds))
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.opts.STSURL+stsTokenPath, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	if c.opts.RequestID != "" {
		req.Header.Set("X-Request-Id", c.opts.RequestID)
	}
	res, err := c.httpClient.Do(req)
	if err != nil {
		return "", &ControlClientError{Stage: "token", Status: 0, Reason: err.Error()}
	}
	defer res.Body.Close()
	parsed := readBody(res.Body)
	if res.StatusCode >= 400 {
		reason, code, remediation := describeError(parsed, "token exchange rejected")
		return "", &ControlClientError{Stage: "token", Status: res.StatusCode, Reason: reason, Code: code, Remediation: remediation}
	}
	envelope, _ := parsed.(map[string]any)
	token, _ := envelope["access_token"].(string)
	if token == "" {
		return "", &ControlClientError{Stage: "token", Status: res.StatusCode, Reason: "token exchange returned no access_token"}
	}
	return token, nil
}

func readBody(body io.Reader) any {
	data, err := io.ReadAll(body)
	if err != nil || len(data) == 0 {
		return nil
	}
	var parsed any
	if json.Unmarshal(data, &parsed) != nil {
		return map[string]any{"raw": string(data)}
	}
	return parsed
}

func describeError(body any, fallback string) (string, string, string) {
	envelope, ok := body.(map[string]any)
	if !ok {
		return fallback, "", ""
	}
	switch wire := envelope["error"].(type) {
	case map[string]any:
		reason, _ := wire["reason"].(string)
		if reason == "" {
			reason = fallback
		}
		code, _ := wire["code"].(string)
		remediation, _ := wire["remediation"].(string)
		return reason, code, remediation
	case string:
		return wire, "", ""
	}
	return fallback, "", ""
}
